compressors/fano_coder.py

```diff
@@ -0,0 +1,154 @@
+"""Fano Coding
+Shannon coding is a basic prefix free coding. There is some confusion in literature about Shannon, Fano and
+Shannon-Fano codes, e.g. see Naming section of https://en.wikipedia.org/wiki/Shannon–Fano_coding.
+
+This document uses Fano coding as described in the wiki article above.
+"""
+from typing import Any, Tuple
+from utils.bitarray_utils import BitArray
+from utils.test_utils import get_random_data_block, try_lossless_compression
+from compressors.prefix_free_compressors import PrefixFreeTree, PrefixFreeEncoder, PrefixFreeDecoder, BinaryNode
+from core.prob_dist import ProbabilityDist
+
+
+class FanoTree(PrefixFreeTree):
+    def __init__(self, prob_dist):
+        super().__init__(prob_dist)
+        # sort the probability distribution
+        self.sorted_prob_dist = ProbabilityDist.get_sorted_prob_dist(prob_dist.prob_dict)
+        # initialize root node of Fano Tree
+        self.root_node = BinaryNode(id=None)
+
+        # build tree returns the root node
+        super().__init__(root_node=self.build_fano_tree(self.root_node, self.sorted_prob_dist))
+
+    @staticmethod
+    def normalize_prob_dict(prob_dict):
+        """
+        normalizes subset of original probability dict to 1
+        """
+        sum_p = sum(prob_dict.values())
+        return ProbabilityDist(dict([[a, b / sum_p] for a, b in prob_dict.items()]))
+
+    @staticmethod
+    def criterion(dict_item):
```

Comment:
Instead of the function "criterion" which doesnt make much logical sense, 
probably move some of the prob_dist logic and create the following func:
```
    def split_prod_dist_into_two(prob_dist):
        # find the halfway split index
        split_index = min(prob_dist.cumulative_prob_dict.values(), key=lambda x: abs(x - 0.5))

        # split prob_dist into two based on split_index
        ...

        # normalize both of the dicts
        return left_prob_dist, right_prob_dist
```